mitodl/mitxpro
# Fix basket sentry errors

#### Pre-Flight checklist


#### What are the relevant tickets?
No ticket, but related to https://github.com/mitodl/mitxpro/issues/843

#### What's this PR do?
- Gets rid of unnecessary Sentry errors that were being raised from the checkout front-end
- Refactors some of the checkout JS code, which had some readability issues

#### How should this be manually tested?
The checkout page should be tested through the "normal" flow (user successfully completes checkout, with or without promo code, and with or without a product id specified in the querystring).

After that, test various error scenarios:
- Basket API endpoint returns an error response without `{"errors": ...}` in the body. This should result in a generic error message (contact customer support)
- Basket API endpoint returns an error response with `{"errors": {"runs": {...}}}` in the body when you select a run. An error message should be shown beneath the dropdown
- Invalid promo code entered. An error message should be shown beneath the input
- Basket API and checkout API endpoints return 403 responses. This should redirect you to the login page. The checkout endpoint is called when you submit, and the basket endpoint is called (a) on page load, (b) when you enter a promo code, and (c) when you select a new run. All of those actions should result in a redirect
- Basket API endpoint returns an error response without `{"errors": ...}` in the body for any request after loading (e.g.: a new promo code is entered or a new run is selected). This should result in a generic error message beneath the promo code input
- Checkout API endpoint returns an error response without `{"errors": ...}` in the body. This should result in a generic error message beneath the submit button after you click it

#### Any background context you want to provide?
Example of unneeded Sentry error: https://sentry.io/organizations/mit-office-of-digital-learning/issues/1987256930/?project=1413655&referrer=slack
